src/lib/consensus/proof_of_stake.ml

```diff
@@ -413,17 +438,29 @@ module Data = struct
           !t.curr_epoch_snapshot <- v
 
     let seen_slot (t : t) epoch slot =
-      match
-        Tuple2.compare ~cmp1:Epoch.compare ~cmp2:Epoch.Slot.compare
-          !t.last_checked_slot_and_epoch (epoch, slot)
-      with
-      | i when i >= 0 ->
-          `Seen
-      | i when i < 0 ->
-          !t.last_checked_slot_and_epoch <- (epoch, slot) ;
-          `Unseen
-      | _ ->
-          failwith "forall x : int. x >= 0 || x < 0 is impossibly false"
+      let module Table = Public_key.Compressed.Table in
+      let unseens =
+        Table.to_alist !t.last_checked_slot_and_epoch
+        |> List.filter_map ~f:(fun (pk, old_epoch_and_slot) ->
+               match
+                 Tuple2.compare ~cmp1:Epoch.compare ~cmp2:Epoch.Slot.compare
+                   old_epoch_and_slot (epoch, slot)
+               with
+               | i when i >= 0 ->
+                   None
+               | i when i < 0 ->
+                   Table.set !t.last_checked_slot_and_epoch ~key:pk
+                     ~data:(epoch, slot) ;
+                   Some pk
+               | _ ->
+                   failwith
+                     "forall x : int. x >= 0 || x < 0 is impossibly false" )
```

Comment:
`if i >= 0 then ... else ...` not good enough? This is a binary match with an impossible default, seems pretty smelly.

Reply:
```
0ff630230 src/lib/consensus/proof_of_stake.ml (Nathan Holland   2019-05-02 15:28:06 -0700  357)       match
0ff630230 src/lib/consensus/proof_of_stake.ml (Nathan Holland   2019-05-02 15:28:06 -0700  358)         Tuple2.compare ~cmp1:E
poch.compare ~cmp2:Epoch.Slot.compare
0ff630230 src/lib/consensus/proof_of_stake.ml (Nathan Holland   2019-05-02 15:28:06 -0700  359)           t.last_checked_slot_
and_epoch (epoch, slot)
0ff630230 src/lib/consensus/proof_of_stake.ml (Nathan Holland   2019-05-02 15:28:06 -0700  360)       with
0ff630230 src/lib/consensus/proof_of_stake.ml (Nathan Holland   2019-05-02 15:28:06 -0700  361)       | i when i >= 0 ->
0ff630230 src/lib/consensus/proof_of_stake.ml (Nathan Holland   2019-05-02 15:28:06 -0700  362)           `Seen
0ff630230 src/lib/consensus/proof_of_stake.ml (Nathan Holland   2019-05-02 15:28:06 -0700  363)       | i when i < 0 ->
0ff630230 src/lib/consensus/proof_of_stake.ml (Nathan Holland   2019-05-02 15:28:06 -0700  364)           t.last_checked_slot_and_epoch <- (epoch, slot) ;
0ff630230 src/lib/consensus/proof_of_stake.ml (Nathan Holland   2019-05-02 15:28:06 -0700  365)           `Unseen
0ff630230 src/lib/consensus/proof_of_stake.ml (Nathan Holland   2019-05-02 15:28:06 -0700  366)       | _ ->
0ff630230 src/lib/consensus/proof_of_stake.ml (Nathan Holland   2019-05-02 15:28:06 -0700  367)           failwith "forall x : int. x >= 0 || x < 0 is impossibly false"
0ff630230 src/lib/consensus/proof_of_stake.ml (Nathan Holland   2019-05-02 15:28:06 -0700  368)   end
```

Reply:
but I'll change it :)

Reply:
I should just give myself code reviews from now on I guess :)